src/middleware/initiate.js

```diff
@@ -1,8 +1,98 @@
 'use strict'
 
 const uuid = require('uuid')
+const {DateTime} = require('luxon')
 
 const logger = require('../logger')
+const {createState, readStateByEndpointId} = require('../db/services/states')
+const {extractValueFromObject, handleServerError} = require('../util')
+
+const extractByType = (type, extract, allData) => {
+  let state = {}
+  Object.keys(extract[type]).forEach(prop => {
+    const value = extractValueFromObject(allData[type], extract[type][prop])
+    state[prop] = value
+  })
+  return state
+}
+
+const extractStateValues = (ctx, extract) => {
+  const allData = ctx.state.allData
+  let updatedState = {}
+
+  // always add timestamps to the endpoint state
+  ctx.state.allData.timestamps.endpointEnd = DateTime.utc().toISO()
+  ctx.state.allData.timestamps.endpointDuration = DateTime.fromISO(
+    ctx.state.allData.timestamps.endpointEnd
+  )
+    .diff(DateTime.fromISO(ctx.state.allData.timestamps.endpointStart))
+    .toObject()
+  updatedState.system = {
+    timestamps: ctx.state.allData.timestamps
+  }
+
+  if (!extract) {
+    // return the default state if no user supplied state defined
+    return updatedState
+  }
+
+  if (extract.requestBody && Object.keys(extract.requestBody).length > 0) {
+    updatedState.requestBody = extractByType('requestBody', extract, allData)
+  }
+
+  if (extract.responseBody && Object.keys(extract.responseBody).length > 0) {
+    updatedState.responseBody = extractByType('responseBody', extract, allData)
+  }
+
+  if (extract.query && Object.keys(extract.query).length > 0) {
+    updatedState.query = extractByType('query', extract, allData)
+  }
+
+  if (
+    extract.lookupRequests &&
+    Object.keys(extract.lookupRequests).length > 0
+  ) {
+    updatedState.lookupRequests = extractByType(
+      'lookupRequests',
+      extract,
+      allData
+    )
+  }
+
+  return updatedState
+}
+
+const updateEndpointState = async (ctx, endpoint) => {
+  if (!endpoint || Object.keys(endpoint).length === 0) {
+    throw new Error('No metaData supplied for updating state for this endpoint')
+  }
+
+  if (!endpoint.state || Object.keys(endpoint).length === 0) {
```

Comment:
@BMartinos Will this case ever exist?

Reply:
not anymore not. this will be removed as there will always be a state object. At minimum with the request timestamps in :+1:

Reply:
This logic has been updated to always be triggered for each request